relays/client-substrate/src/client/caching.rs

```diff
@@ -175,41 +183,71 @@ impl<C: Chain, B: Client<C>> Client<C> for CachingClient<C, B> {
 	}
 
 	async fn best_finalized_header_hash(&self) -> Result<HashOf<C>> {
-		// TODO: after https://github.com/paritytech/parity-bridges-common/issues/2074 we may
-		// use single-value-cache here, but for now let's just call the backend
-		self.backend.best_finalized_header_hash().await
+		let (result_sender, result_receiver) = bounded(1);
+		self.data
+			.to_background_sender
+			.read()
+			.await
+			.try_send(ToBackground::BestFinalizedHeader(result_sender))?;
+		let best_finalized_header = result_receiver.recv().await?;
+
+		// update cache
+		let best_finalized_header_hash = best_finalized_header.hash();
+		self.data
+			.header_by_hash_cache
+			.write()
+			.await
+			.insert(best_finalized_header_hash, best_finalized_header);
+		Ok(best_finalized_header_hash)
 	}
 
 	async fn best_header(&self) -> Result<HeaderOf<C>> {
-		// TODO: if after https://github.com/paritytech/parity-bridges-common/issues/2074 we'll
-		// be using subscriptions to get best blocks, we may use single-value-cache here, but for
-		// now let's just call the backend
-		self.backend.best_header().await
+		let (result_sender, result_receiver) = bounded(1);
```

Comment:
It seems a bit complicated to send a command to the worker for anything that we need. Could we hold the data that is updated by the background worker in shared memory (a separate structure) ? For example something like:

```
struct BackgroundWorkerData<C: Chain> {
	best_header: RwLock<Option<C::Header>>,
    ...
}

impl<C: Chain> BackgroundWorkerData<C> {
	async fn best_header<B: Client<C>>(&self, backend: &B) -> Result<C::Header> {
		Ok(self.best_header.read().await.clone().unwrap_or(backend.best_header().await?))
	}

    ...
}
```

The backround worker would write to it, and the client would read from it, making it easier to access the values. For example:
```
	async fn best_header(&self) -> Result<HeaderOf<C>> {
		self.worker_data.best_header(&self.backend).await
	}
```

Reply:
I agree - `RwLock`s looks better here :) I had a sync-primitives allergy after investigating this recent relay-halt-issue, hence the channels. Fixed

Reply:
@serban300 I was considering another your suggestion (below) and realized that there was another reason to do that using channels. For example if background task has exited (because it has encountered some error), and if we are using `Arc<RwLock<>>`, then there's no way for us to understand that. And if we keep using channels, we could understand that and report an error to the caller. So I'm thinking of reverting my last commit (https://github.com/paritytech/parity-bridges-common/pull/2166/commits/ff6af52cbf53b51b86284c849c47cd05a8b0ef46).

We can also start background task for processing the subscriptions and change type of `best_header` from `Arc<RwLock<Option<_>>>` and `Arc<RwLock<Option<Result<_, _>>>>`, but imo channels are more reliable. WDYT?

Reply:
We could also have a `background_worker_is_running` flag maybe ? And if it's false, we don't return from the cache. For best_header and best_finalized_header.

Reply:
Or just clear the caches if the background worker task exits ?

Reply:
My point is that relying on the fact that background task will set something **before** it stops doesn't feels like a super-reliable solution. I could just change the types from `Arc<RwLock<Option<_>>>` to `Arc<RwLock<Option<Result<_, _>>>>` and set it to `Some(Err())` before it exits. Do you think it is better than reverting to channels?

Reply:
I think there should be a reliable way to perform some logic after the background task closes. For example `async_std::task::spawn()` returns a `JoinHandle`. Maybe we can poll that. If there is any result, it should mean that the background task exited ? I will do some experiments to see if that's true (to see for example if this works correctly if the task is killed) and if it's usable. Will get back with more details.

But I wouldn't use `Arc<RwLock<Option<Result<_, _>>>>` . It adds more complexity. I would just clear the caches.

Reply:
Looked a bit into it. The only way I could find to break the async task and prevent it from running the logic after `do_background_worker()` was to have a panic inside `do_background_worker`. To handle this we could have a guard struct that is moved to the background worker task when it is spawned and that implements `Drop`. and on drop it performs the cache cleanup logic. The problem is that drop is not async, so we would have to use standard RwLocks or Mutexes if possible. 

Another option would be to have just a Ping command and send it through a channel before querying the cache. If the ping fails, we don't use the cache.

Anyway, maybe more research would be needed and it's not worth blocking the task. For me it's ok to revert the last commit as you suggested, and merge it like this for the moment. And after that I can do some more research about the guard solution or the ping or others.

Reply:
Let's sleep over it - I don't think there's anything better than just call `receiver.recv()?`. Everything else is more complicated, so probably channels are the way to go. But let's reconsider on next week. Anyway - thanks for review and for suggestions :)